File: hibernate-scala-examples/src/main/scala/org/hibernate/examples/hibernate/interceptor/PersistentObjectInterceptor.scala
package org.hibernate.examples.interceptor

import java.io.Serializable
import org.hibernate.EmptyInterceptor
import org.hibernate.`type`.Type
import org.hibernate.examples.model.PersistentObject

/**
 * kr.debop4s.data.hibernate.interceptor.PersistentObjectInterceptor
 *
 * @author 배성혁 sunghyouk.bae@gmail.com
 * @since 2013. 12. 24. 오후 2:38
 */
class PersistentObjectInterceptor extends EmptyInterceptor {

    def isPersisted(entity: AnyRef): Boolean = entity match {
        case p: PersistentObject => true
        case _ => false
    }

    override
    def onLoad(entity: Any, id: Serializable, state: Array[AnyRef], propertyNames: Array[String], types: Array[Type]): Boolean = {
        entity match {
            case p: PersistentObject => p.onLoad()
            case _ =>
        }
        false
    }

    override
    def onSave(entity: scala.Any, id: Serializable, state: Array[AnyRef], propertyNames: Array[String], types: Array[Type]): Boolean = {
        entity match {
            case p: PersistentObject => p.onPersist()
            case _ =>
        }
        false
    }
}

File: hibernate-examples/src/main/java/org/hibernate/examples/hibernate/interceptor/PersistentObjectInterceptor.java
package org.hibernate.examples.hibernate.interceptor;

import lombok.extern.slf4j.Slf4j;
import org.hibernate.EmptyInterceptor;
import org.hibernate.examples.model.PersistentObject;

import java.io.Serializable;

/**
 * Hibernate PersistentObject의 영구저장 여부를 관리할 수 있도록 해주는 Interceptor입니다.
 *
 * @author 배성혁 sunghyouk.bae@gmail.com
 * @since 2013. 11. 28. 오전 11:06
 */
@Slf4j
public class PersistentObjectInterceptor extends EmptyInterceptor {

  public boolean isPersisted(Object entity) {
    return entity instanceof PersistentObject && ((PersistentObject) entity).isPersisted();
  }

  /**
   * 엔티티 로드 후 {@link org.hibernate.examples.model.PersistentObject#onLoad()} 를 호출합니다.
   */
  @Override
  public boolean onLoad(Object entity,
                        Serializable id,
                        Object[] state,
                        String[] propertyNames,
                        org.hibernate.type.Type[] types) {
    if (entity instanceof PersistentObject) {
      ((PersistentObject) entity).onLoad();
    }
    return false;
  }

  /**
   * 엔티티 저장 후 {@link org.hibernate.examples.model.PersistentObject#onSave()}를 호출합니다.
   */
  @Override
  public boolean onSave(Object entity,
                        Serializable id,
                        Object[] state,
                        String[] propertyNames,
                        org.hibernate.type.Type[] types) {
    if (entity instanceof PersistentObject) {
      ((PersistentObject) entity).onSave();
    }
    return false;
  }

  private static final long serialVersionUID = -9072457259066614636L;
}
